File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/distanceprocessors/DistanceProcessor.java
package co.electric.snake.strategy.bonostrategy.distanceprocessors;

import co.electric.snake.framework.model.Coordinate;
import co.electric.snake.framework.model.Direction;

public abstract class DistanceProcessor {

    public static DistanceProcessor getStrategy(Direction direction) {
        switch (direction) {
            case NORTH:
                return new NorthLookingDistanceStrategy();
            case SOUTH:
                return new SouthLookingDistanceStrategy();
            case EAST:
                return new EastLookingDistanceStrategy();
            case WEST:
                return new WestLookingDistanceStrategy();
            default:
                return new NorthLookingDistanceStrategy();
        }
    }

    public abstract int getDistance(Coordinate actualCoordinate,
                                    Coordinate blockingCoordinate, Coordinate maxCoordinate);

    protected int getDistanceDirectly(int biggerCoordinateFactor,
                                      int smallerCoordinateFactor) {
        return biggerCoordinateFactor - smallerCoordinateFactor;
    }

    protected int getDistanceClipped(int directCoordinateFactor,
                                     int clipperCoordinateFactor, int maxCoordinateFactor) {
        return directCoordinateFactor
                + (maxCoordinateFactor - clipperCoordinateFactor);
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/directionprocessor/distanceprocessor/DistanceProcessor.kt
package co.electric.snake.strategy.bonostrategy.directionprocessor.distanceprocessor

import co.electric.snake.framework.model.Coordinate
import co.electric.snake.framework.model.Direction
import java.util.*

interface DistanceProcessor {

    fun getDistance(direction: Direction, headCoordinate: Coordinate, blockingCoordinate: Coordinate, maxCoordinate: Coordinate): Optional<Int>

    fun getDistanceDirectly(biggerCoordinateFactor: Int, smallerCoordinateFactor: Int): Int {
        return biggerCoordinateFactor - smallerCoordinateFactor
    }

    fun getDistanceClipped(directCoordinateFactor: Int, clipperCoordinateFactor: Int, maxCoordinateFactor: Int): Int {
        return (directCoordinateFactor + (maxCoordinateFactor - clipperCoordinateFactor))
    }

}
